### PR TITLE
 feature: change to zipdeploy & run from package

### PR DESCRIPTION
**Pull request by mydiemho**
1. use the current Azure recommended zip deploy method to [run function from deployment package ](https://docs.microsoft.com/en-us/azure/azure-functions/run-functions-from-deployment-package)
1. remove support for individual function deploy since that's not supported by zipdeploy

### CODE REVIEW

File: src/config.ts [hunk 1]
```diff
@@ -13,7 +13,8 @@ export const constants = {
   scmCommandApiPath: '/api/command',
   scmDomain: '.scm.azurewebsites.net',
   scmVfsPath: '/api/vfs/site/wwwroot/',
-  scmZipApiPath: '/api/zip/site/wwwroot/'
+  scmZipApiPath: '/api/zip/site/wwwroot/',
```
Comment (pjlittle) [inline src/config.ts]: Do we still need to have the old endpoint const?

File: src/services/functionAppService.ts [hunk 2]
```diff
@@ -48,68 +49,34 @@ export class FunctionAppService extends BaseService {
...
-        method: 'PUT',
-        uri: `https://${scmDomain}/api/zip/site/wwwroot/${functionName}/`,
+        method: 'POST',
+        uri: uploadUrl,
         json: true,
         headers: {
           Authorization: `Bearer ${this.credentials.tokenCache._entries[0].accessToken}`,
-          Accept: '*/*'
+          Accept: '*/*',
+          "Content-type": "application/octet-stream"
```
Comment (pjlittle) [inline src/services/functionAppService.ts]: not sure if it matters to Azure, but typically multi-word headers are camel case: `Content-Type`

nitpic: Different than TypeScript, but it looks like the convention is to use single quotes vs. double quotes.

File: src/services/functionAppService.ts [hunk 3]
```diff
@@ -48,68 +49,34 @@ export class FunctionAppService extends BaseService {
...
         }
       };
 
       await this._sendFile(requestOptions, functionZipFile);
```
Comment (pjlittle) [inline src/services/functionAppService.ts]: Feels like we need some error handling/logging for when this fails.  I'm also curious on timeout... if you have a big enough file, will it kill the upload? Or maybe all of this is covered in the `_sendFile` function?

**Comment by pjlittle** [comment]
Looking good. Does it make sense to add corresponding tests or are we going to revisit that for the deeper refactor efforts?

File: src/services/functionAppService.ts (hunk 3 above)
Comment (mydiemho) [reply src/services/functionAppService.ts]: I was wondering too what’s the best way to handle this.  _sendFile is an async function already but the zipdeploy api also let you set an isAsyncFlag (https://github.com/projectkudu/kudu/wiki/Deploying-from-a-zip-file-or-url#asynchronous-zip-deployment).  Though I’m not sure how an async within an async would work.

**Comment by mydiemho** [comment]
@pjlittle we should discuss how we want to approach unit tests at standup

File: src/config.ts (hunk 1 above)
Comment (mydiemho) [reply src/config.ts]: Good catch, we can remove it

File: src/services/functionAppService.ts (hunk 3 above)
Comment (tbarlow12) [reply src/services/functionAppService.ts]: I think even a simple try/catch around the `await` that logs the error could be a start

File: src/plugins/deploy/azureDeployPlugin.ts [hunk 4]
```diff
@@ -16,19 +16,18 @@ export class AzureDeployPlugin {
...
   private async deploy() {
     const resourceService = new ResourceService(this.serverless, this.options);
     await resourceService.deployResourceGroup();
 
     const functionAppService = new FunctionAppService(this.serverless, this.options);
-    const functionApp = await functionAppService.deploy();
 
-    await functionAppService.uploadFunctions(functionApp);
-    await functionAppService.syncTriggers(functionApp);
+    // create all necessary resources: resource-group, storage account, app service plan, and app service
```
Comment (tbarlow12) [inline src/plugins/deploy/azureDeployPlugin.ts]: Is this comment saying that line 30 does all that? Or that this is where we _would_ do that?

Comment (mydiemho) [reply src/plugins/deploy/azureDeployPlugin.ts]: this is saying line 30 does all that

File: src/plugins/deploy/azureDeployPlugin.ts [hunk 5]
```diff
@@ -13,22 +13,17 @@ export class AzureDeployPlugin {
   }
 
   private async beforeDeploy() {
```
Comment (wbreza) [inline src/plugins/deploy/azureDeployPlugin.ts]: Let's remove the whole hook till we have something relevant to put here.

File: src/services/functionAppService.ts [hunk 6]
```diff
@@ -43,63 +44,45 @@ export class FunctionAppService extends BaseService {
...
-  public async uploadFunctions(functionApp): Promise<any> {
-    this.serverless.cli.log('Creating azure functions');
-
-    const scmDomain = functionApp.enabledHostNames[0];
+    // Upload function artifact if it exists, otherwise the full service is handled in 'uploadFunctions' method
     const functionZipFile = this.serverless.service['artifact'];
+    if (functionZipFile) {
+      this.serverless.cli.log(`-> Uploading ${functionZipFile}`);
+
+      const uploadUrl = `https://${functionAppName}${constants.scmDomain}${constants.scmZipDeployApiPath}`;
```
Comment (wbreza) [inline src/services/functionAppService.ts]: I don't think these constants should be used.  In theory this should be the only function that should use this URL.  Also - the `scmDomain` is not a constant and will change based on the configuration of the function app.

File: src/services/functionAppService.ts [hunk 7]
```diff
@@ -43,63 +44,45 @@ export class FunctionAppService extends BaseService {
     return response.data.value;
   }
 
-  public async deleteFunction(functionName) {
```
Comment (wbreza) [inline src/services/functionAppService.ts]: We don't know if we will need `syncTriggers` or `deleteFucntion` in some other workflows.  Let's keep these methods for now.

File: src/services/functionAppService.ts [hunk 8]
```diff
@@ -43,63 +44,45 @@ export class FunctionAppService extends BaseService {
...
-    this.serverless.cli.log('Syncing function triggers');
-
-    const syncTriggersUrl = `${this.baseUrl}${functionApp.id}/syncfunctiontriggers?api-version=2016-08-01`;
-    await this.sendApiRequest('POST', syncTriggersUrl);
-  }
+  async zipDeploy(functionApp) {
+    const functionAppName = functionApp.name;
+    this.serverless.cli.log(`Deploying zip file to function app: ${functionAppName}`);
 
-  public async cleanUp(functionApp) {
```
Comment (wbreza) [inline src/services/functionAppService.ts]: Same with `cleanUp` and `listFunctions`.  We may need these in other workflows.

File: src/services/functionAppService.ts [hunk 9]
```diff
@@ -43,63 +44,45 @@ export class FunctionAppService extends BaseService {
...
-  }
-
-  public async listFunctions(functionApp) {
-    const getTokenUrl = `${this.baseUrl}${functionApp.id}/functions?api-version=2016-08-01`;
-    const response = await this.sendApiRequest('GET', getTokenUrl);
-
-    return response.data.value || [];
-  }
-
-  public async uploadFunctions(functionApp): Promise<any> {
```
Comment (wbreza) [inline src/services/functionAppService.ts]: I know we are using `zipDeploy` method for the service deployment - but I think we should keep a more generic name as the method name.  Later on when we introduce `runFromPackage` we will have multiple branches within this method.  Let's stick with `uploadFunctions` for now.

File: src/services/functionAppService.ts [hunk 10]
```diff
@@ -43,63 +44,45 @@ export class FunctionAppService extends BaseService {
...
-    this.serverless.cli.log('Creating azure functions');
-
-    const scmDomain = functionApp.enabledHostNames[0];
+    // Upload function artifact if it exists, otherwise the full service is handled in 'uploadFunctions' method
     const functionZipFile = this.serverless.service['artifact'];
+    if (functionZipFile) {
```
Comment (wbreza) [inline src/services/functionAppService.ts]: Invert if statement to reduce nesting.

File: src/services/functionAppService.ts (hunk 6 above)
Comment (mydiemho) [reply src/services/functionAppService.ts]: `scmDomain` is coming from `config.ts`, how will that change with the function app?

File: src/services/functionAppService.ts (hunk 9 above)
Comment (mydiemho) [reply src/services/functionAppService.ts]: just to be clear, I am adding `runFromPackage` in this PR already, it's the change in the arm template

File: src/provider/armTemplates/azuredeploy.json [hunk 11]
```diff
@@ -103,6 +103,10 @@
                         {
                             "name": "WEBSITE_NODE_DEFAULT_VERSION",
                             "value": "8.11.1"
```
Comment (pjlittle) [inline src/provider/armTemplates/azuredeploy.json]: Just occurred to me that we should probably target the latest Node 8.x LTX, v8.16.0, which comes with NPM v.4.1: https://nodejs.org/en/download/releases/

I would assume this is supported?

**Comment by pjlittle** [comment]
LGTM :shipit:

File: src/provider/armTemplates/azuredeploy.json (hunk 11 above)
Comment (mydiemho) [reply src/provider/armTemplates/azuredeploy.json]: going to put this in a different pr and test :D

**Pull request merged**